test/e2e/workspace/deletion_test.go

```diff
@@ -0,0 +1,136 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workspace
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	logicalclusterdeletion "github.com/kcp-dev/kcp/pkg/reconciler/core/logicalclusterdeletion/deletion"
+	"github.com/kcp-dev/kcp/sdk/apis/core"
+	corev1alpha1 "github.com/kcp-dev/kcp/sdk/apis/core/v1alpha1"
+	kcpclientset "github.com/kcp-dev/kcp/sdk/client/clientset/versioned/cluster"
+	kcptesting "github.com/kcp-dev/kcp/sdk/testing"
+	kcptestinghelpers "github.com/kcp-dev/kcp/sdk/testing/helpers"
+	"github.com/kcp-dev/kcp/test/e2e/framework"
+)
+
+// TestWorkspaceLogicalClusterRelationship tests that a workspace gets not
+// deleted prematurely if its LogicalCluster has not yet been deleted.
+func TestWorkspaceLogicalClusterRelationship(t *testing.T) {
+	t.Parallel()
+	framework.Suite(t, "control-plane")
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	t.Cleanup(cancelFunc)
+
+	server := kcptesting.SharedKcpServer(t)
+	cfg := server.BaseConfig(t)
+
+	// This will create structure as below for testing:
+	// └── root
+	//   └── e2e-workspace-n784f
+	//    └── test
+	wsName := "test"
+	fixtureRoot, _ := kcptesting.NewWorkspaceFixture(t, server, core.RootCluster.Path())
+	testPath, _ := kcptesting.NewWorkspaceFixture(t, server, fixtureRoot, kcptesting.WithName(wsName))
+
+	clientset, err := kcpclientset.NewForConfig(cfg)
+	require.NoError(t, err, "error creating kube cluster client set")
+
+	lc, err := clientset.Cluster(testPath).CoreV1alpha1().LogicalClusters().Get(ctx, corev1alpha1.LogicalClusterName, v1.GetOptions{})
+	require.NoError(t, err, "error getting logicalcluster")
+
+	// add a finalizer to the cluster, mimicking an external finalizer
+	customFinalizer := "example.com/test"
+	require.EventuallyWithT(t, func(c *assert.CollectT) {
+		lcsUpd := lc.DeepCopy()
+		lcsUpd.Finalizers = append(lcsUpd.Finalizers, customFinalizer)
+		_, err = clientset.Cluster(testPath).CoreV1alpha1().LogicalClusters().Update(ctx, lcsUpd, v1.UpdateOptions{})
+		require.NoError(c, err, "error updating logicalcluster")
+
+		// delete the workspace
+		err = clientset.Cluster(fixtureRoot).TenancyV1alpha1().Workspaces().Delete(ctx, wsName, v1.DeleteOptions{})
+		require.NoError(c, err, "error deleting workspace")
+	}, wait.ForeverTestTimeout, 100*time.Millisecond, "waiting for finalizer to be applied and workspace to be deleted")
```

Comment:
These two things should be in separate `Eventually` wraps, they are two distinct operations that can fail.

Can we also (here and in other places) use `kcptestinghelpers.Eventually`? It starts logging the condition if it's not met after some time, plus it's used later in this code as well, and it would be good to use this consistently.

I think the latter part is on me, I said `require.Eventually` in the previous review. Apologies for that :/